extensions/common/node_info_cache.cc

```diff
@@ -0,0 +1,91 @@
+/* Copyright 2019 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "extensions/common/node_info_cache.h"
+#include "extensions/common/context.h"
+
+#ifdef NULL_PLUGIN
+
+using Envoy::Extensions::Common::Wasm::Null::Plugin::getStringValue;
+using Envoy::Extensions::Common::Wasm::Null::Plugin::getStructValue;
+using Envoy::Extensions::Common::Wasm::Null::Plugin::logDebug;
+using Envoy::Extensions::Common::Wasm::Null::Plugin::logInfo;
+using google::protobuf::util::Status;
+
+#endif  // NULL_PLUGIN
+
+namespace Wasm {
+namespace Common {
+
+namespace {
+
+// getNodeInfo fetches peer node info from host filter state. It returns true if
+// no error occurs.
+bool getNodeInfo(StringView peer_metadata_key,
+                 wasm::common::NodeInfo* node_info) {
+  google::protobuf::Struct metadata;
+  if (!getStructValue({"filter_state", peer_metadata_key}, &metadata)) {
+    LOG_DEBUG(absl::StrCat("cannot get metadata for: ", peer_metadata_key));
+    return false;
+  }
+
+  auto status = ::Wasm::Common::extractNodeMetadata(metadata, node_info);
+  if (status != Status::OK) {
+    LOG_DEBUG(absl::StrCat("cannot parse peer node metadata ",
+                           metadata.DebugString(), ": ", status.ToString()));
+    return false;
+  }
+  return true;
+}
+
+}  // namespace
+
+NodeInfoPtr NodeInfoCache::getPeerById(StringView peer_metadata_id_key,
+                                       StringView peer_metadata_key) {
+  if (max_cache_size_ < 0) {
+    // Cache is disabled, fetch node info from host.
+    auto node_info_ptr = std::make_shared<wasm::common::NodeInfo>();
+    if (getNodeInfo(peer_metadata_key, node_info_ptr.get())) {
+      return node_info_ptr;
+    }
+    return nullptr;
+  }
+
+  std::string peer_id;
+  if (!getStringValue({"filter_state", peer_metadata_id_key}, &peer_id)) {
+    LOG_DEBUG(absl::StrCat("cannot get metadata for: ", peer_metadata_id_key));
+    return nullptr;
+  }
+  auto nodeinfo_it = cache_.find(peer_id);
+  if (nodeinfo_it != cache_.end()) {
+    return nodeinfo_it->second;
+  }
+
+  // Do not let the cache grow beyond max_cache_size_.
+  if (int32_t(cache_.size()) > max_cache_size_) {
+    auto it = cache_.begin();
+    cache_.erase(cache_.begin(), std::next(it, max_cache_size_ / 4));
```

Comment:
Just to confirm my understanding, the cache revocation strategy is to remove 1/4th of the elements at random?

Reply:
Yes, there cache is not sophisticated enough. We should move to lru. @mandarjog wdyt?

Reply:
Sure lru works. But this is a trivial implementation...

Reply:
Yeah, let's not over-complicate until we can measure the effects first.
Can you add a test that exercises cache eviction? I don't see it in this PR.

Reply:
Yeah, definitely not in this PR or 1.4 release, just something we could think about.